### ISSUE TITLE
Most hypervisors are broken when ganeti-masterd is not run as root

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#372]
Originally reported of Google Code with ID 316.

```
EnsureDirs is called by __init__ in classes:
FakeHypervisor
ChrootManager
KVMHypervisor
LXCHypervisor

On the other hand masterd in various functions initializes those classes.
If masterd is not run as root, EnsureDirs fails because masterd doesn't have enough permissions.

Possible solutions:
- Ignore EPERM
- Check if uid=NODED_USER and otherwise skip the call to EnsureDirs
- Remove call to EnsureDirs and move it to somewhere else (which is called only from noded)
- Pass a special parameter when initializing an hypervisor from masterd and in that case skip the call

Other ideas?
```
Originally added on 2012-11-19 12:49:42 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I think you should move the creation of those directories to ensure-dirs (lib/tools/ensure_dirs.py). It runs as root.
```
Originally added on 2012-11-19 12:52:16 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
In the original design, masterd should not initialise the class, just use class methods. So __init__ should be run only on noded by default.

What are we doing that requires masterd initialising the class?
```
Originally added on 2012-11-19 12:52:45 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
We are calling "GetHypervisor" rather than "GetHypervisorClass" in multiple places of cmdlib.py. We can change those calls and hope nothing else breaks, then?
```
Originally added on 2012-11-19 13:01:59 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
hansmi: ensure-dirs is called only at daemon start time: would it be ok then to have directories for "all" hypervisors, even disabled ones, being always created/ensured?
Otherwise commands like gnt-cluster modify --enabled-hypervisors=... won't work.

Thanks,

Guido
```
Originally added on 2012-11-19 13:06:19 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Yes, I think that's fine. Directories have a near-zero cost.
```
Originally added on 2012-11-19 13:17:36 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-01-16 13:12:46 +0000 UTC.

Added Labels: Type-Defect
Changed State: Accepted
Added to Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
This has been fixed in 2.7:

12f647f cmdlib.py: don't use GetHypervisor

```
Originally added on 2013-02-26 16:01:19 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
This has been fixed in 2.7:

12f647f cmdlib.py: don't use GetHypervisor

```
Originally added on 2013-02-26 16:01:19 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking 2.7 bugs fixed, as 2.7.0~rc1 was released, and should be ready to be tested in production environments.
```
Originally added on 2013-05-09 05:42:48 +0000 UTC.

Changed State: Released
